src/gui/plugins/scene3d/Scene3D.cc

```diff
@@ -2299,8 +2472,21 @@ void TextureNode::PrepareNode()
 
     // This will notify the rendering thread that the texture is now being
     // rendered and it can start rendering to the other one.
-    emit TextureInUse();
+    emit TextureInUse(&this->renderSync);
+  }
+  else
+  {
+    // This is necessary because we're forcing serialization of
+    // of worker and Qt threads via renderSync; and if PrepareNode gets
+    // called twice in a row with the worker thread still finishing the
+    // 1st iteration, it may result in a deadlock for newer versions of Qt;
+    // as WaitForWorkerThread will be called with no corresponding
+    // WaitForQtThreadAndBlock as the worker thread thinks there is
+    // no more jobs to do.
+    emit TextureInUse(&this->renderSync);
```

Comment:
I'll explain the 'rationale', perhaps we can make it more obvious for future maintainers:

The original code did not have TextureInUse outside the if. And it shouldn't. But the original code did not synchronize with the worker thread at all.

Because we're forcing serialization, it is safe to emit the TextureInUse outside the if/else, and is required otherwise we may hit a deadlock (not theoretical, I was hitting it)

We want to leave the door open for not forcing serialization in the future.
If that happens, that emit in the 'else' body should not happen.

Although after some consideration, I believe the code would have to be changed more drastically, in a way that resembles Qt's original code less (i.e. without using signals and slots at all), and uses a custom loop we can synchronize(*) without danger of deadlocks.

(*) The main problem is that Qt's original code did not have to synchronize at all because it just rendered a 512x512 FBO. But in a real world application, we need to synchronize (halt everything) when we must change resolutions due to window resize events.

So what I'm trying to say is:

 - Look, this emit is currently required
 - But if you stop forcing serialization and a allow concurrent rendering using a double buffer scheme, this emit should not be here

Suggestions on how to make this crystal clear in the comments is welcome